In [46]:
import pandas as pd
import numpy as np
import requests

In [3]:
wiki_url = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
wiki_url

<Response [200]>

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

We have 103 rows and 3 columns

# Question 2

Execution doesn't stop so i will importing the CSV file from the URL

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.
Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [18]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [23]:
from geopy.geocoders import Nominatim
address = 'Toronto, Ontario'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [25]:
import folium

In [26]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [27]:
CLIENT_ID = 'S5SAICXPSOGKAUNY0GT4AKAZIQCLZ5ODL412BN1DPR5O02J4' 
CLIENT_SECRET = 'V4X0LDB5DTQAKKBCB0JY2A1YQYYDT202JQSZJCZMSOAO42AQ'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: S5SAICXPSOGKAUNY0GT4AKAZIQCLZ5ODL412BN1DPR5O02J4
CLIENT_SECRET:V4X0LDB5DTQAKKBCB0JY2A1YQYYDT202JQSZJCZMSOAO42AQ


I create a function to get all the venue categories in Toronto

In [29]:
def Neighbourhood(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [37]:
neighbors_in_toronto = Neighbourhood(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Checking the Venues based on Neighbourhood

In [39]:
neighbors_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1311  Mimico NW, The Queensway West, South of Bloor,...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1311               43.628841               -79.520999              Wingporium   
1312               43.628841               -79.520999        South St. Burger   
1313               43.628841               -79.520999               Dollarama   
1314               43.628841               -79.520999          Healthy Planet   
1315               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1311            Wings Joint  
1312           Burger Joint  
1313         Discount Store  
1314        Supplement Shop  
1315                 Bakery  

[414 rows x 5 columns]

# One Hot encoding the venue Categories

In [40]:
toronto_venue_cat = pd.get_dummies(neighbors_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1319                  0        0                   0             0   
1320                  0        0                   0             0   
1321                  0        0                   0             0   
1322                  0        0                   0             0   
1323                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1319               0                0                 0                    0   
1320               0                0                 0                    0   
1321               0                0                 0                    0   
1322               0                0                 0                    0   
1323               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  Truck Stop  \
0                0         0  ...              0           0   
1                0         0  ...              0           0   
2                0         0  ...              0           0   
3                0         0  ...              0           0   
4                0         0  ...              0           0   
...            ...       ...  ...            ...         ...   
1319             0         0  ...              0           0   
1320             0         0  ...              0           0   
1321             0         0  ...              0           0   
1322             0         0  ...              0           0   
1323             0         0  ...              0           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1319                              0                 0                      0   
1320                              0                 0                      0   
1321                              0                 0                      0   
1322                              0                 0                      0   
1323                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0         0            0              0            0  
...               ...   

Adding the neighbourhood to the encoded dataframe

In [43]:
toronto_venue_cat['Neighbourhood'] = neighbors_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Truck Stop  \
0                    0             0  ...              0           0   
1                    0             0  ...              0           0   
2                    0             0  ...              0           0   
3                    0             0  ...              0           0   
4                    0             0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 235 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [44]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.047619           0.0  ...            0.0   

   Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0         0.0                            0.0               0.0   
1         0.0                            0.0               0.0   
2         0.0                            0.0               0.0   
3         0.0                            0.0               0.0   
4         0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 235 columns]

Let's make a function to get the top most common venue categories.

In [45]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [48]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot         Chinese Restaurant                Lounge   
1           Coffee Shop                        Gym          Skating Rink   
2                  Bank  Middle Eastern Restaurant    Chinese Restaurant   
3    Chinese Restaurant                       Bank   Japanese Restaurant   
4           Coffee Shop         Italian Restaurant         Grocery Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Yoga Studio          Dessert Shop           Escape Room   
1        Sandwich Place                   Pub           Yoga Studio   
2           Bridal Shop        Sandwich Place                 Diner   
3      Department Store           Escape Room     Electronics Store   
4       Thai Restaurant             Juice Bar     Indian Restaurant   

         8th Most Common Venue        9th Most Common Venue  \
0            Electronics Store  Eastern European Restaurant   
1           Dim Sum Restaurant             Department Store   
2                   Restaurant                 Intersection   
3  Eastern European Restaurant                    Drugstore   
4                   Restaurant                      Butcher   

  10th Most Common Venue  
0              Drugstore  
1           Dessert Shop  
2          Deli / Bodega  
3             Donut Shop  
4       Sushi Restaurant

Let's make the model to cluster our Neighbourhoods

In [49]:
# import k-means from clustering stage
from sklearn.cluster import KMeans
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Let's add the clustering Label column to the top 10 common venue categories

In [50]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [51]:
toronto_merged = combined_data
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')
toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0  43.753259 -79.329656             1.0                   Park   
1  43.725882 -79.315572             0.0  Portuguese Restaurant   
2  43.654260 -79.360636             0.0            Coffee Shop   
3  43.718518 -79.464763             0.0         Clothing Store   
4  43.662301 -79.389494             0.0            Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio           Deli / Bodega   
1          Hockey Arena          Intersection       French Restaurant   
2                  Park                Bakery          Breakfast Spot   
3     Accessories Store              Boutique  Furniture / Home Store   
4      Sushi Restaurant                   Gym               Nightclub   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant             Drugstore   
1           Coffee Shop                    Drugstore            Donut Shop   
2                  Café                   Restaurant                   Pub   
3           Event Space                  Coffee Shop         Women's Store   
4              Beer Bar                Smoothie Shop        Sandwich Place   

   8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0             Donut Shop                Dog Run    Distribution Center  
1                Dog Run    Distribution Center          Deli / Bodega  
2         Chocolate Shop  Performing Arts Venue           Dessert Shop  
3  Vietnamese Restaurant        Airport Service     Dim Sum Restaurant  
4          Burrito Place                   Café  Portuguese Restaurant

Drop all the NaN values to prevent data skew

In [52]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plot

In [54]:
import matplotlib.cm as cm
import matplotlib.colors as colors

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Cluster 1

In [55]:

toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
1          North York             0.0  Portuguese Restaurant   
2    Downtown Toronto             0.0            Coffee Shop   
3          North York             0.0         Clothing Store   
4    Downtown Toronto             0.0            Coffee Shop   
7          North York             0.0                    Gym   
..                ...             ...                    ...   
97   Downtown Toronto             0.0             Restaurant   
99   Downtown Toronto             0.0            Coffee Shop   
100      East Toronto             0.0     Light Rail Station   
101         Etobicoke             0.0         Baseball Field   
102         Etobicoke             0.0          Grocery Store   

    2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
1            Hockey Arena          Intersection       French Restaurant   
2                    Park                Bakery          Breakfast Spot   
3       Accessories Store              Boutique  Furniture / Home Store   
4        Sushi Restaurant                   Gym               Nightclub   
7              Beer Store           Coffee Shop     Japanese Restaurant   
..                    ...                   ...                     ...   
97                   Café           Coffee Shop      Seafood Restaurant   
99                   Park                 Diner    Ethiopian Restaurant   
100           Yoga Studio         Auto Workshop              Comic Shop   
101         Deli / Bodega           Yoga Studio            Dessert Shop   
102          Burger Joint           Wings Joint           Tanning Salon   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1             Coffee Shop             Drugstore            Donut Shop   
2                    Café            Restaurant                   Pub   
3             Event Space           Coffee Shop         Women's Store   
4                Beer Bar         Smoothie Shop        Sandwich Place   
7              Restaurant        Clothing Store                  Café   
..                    ...                   ...                   ...   
97            Art Gallery                 Hotel            Steakhouse   
99               Beer Bar           Escape Room   Japanese Restaurant   
100           Pizza Place            Restaurant         Burrito Place   
101  Ethiopian Restaurant           Escape Room     Electronics Store   
102        Discount Store                   Gym       Supplement Shop   

           8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
1                        Dog Run    Distribution Center          Deli / Bodega  
2                 Chocolate Shop  Performing Arts Venue           Dessert Shop  
3          Vietnamese Restaurant        Airport Service     Dim Sum Restaurant  
4                  Burrito Place                   Café  Portuguese Restaurant  
7                 Sandwich Place              Bike Shop               Bus Line  
..                           ...                    ...                    ...  
97                        Bakery             Food Court                    Pub  
99                     Bookstore         Breakfast Spot        Bubble Tea Shop  
100                      Brewery             Skate Park         Farmers Market  
101  Eastern European Restaurant              Drugstore             Donut Shop  
102         Fast Food Restaurant      Convenience Store         Hardware Store  

[85 rows x 12 columns]

Cluster 2

In [56]:

toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             1.0                  Park   
16              York             1.0                 Trail   
21              York             1.0                  Park   
35         East York             1.0           Coffee Shop   
49        North York             1.0                  Park   
52        North York             1.0                  Park   
66        North York             1.0                  Park   
83   Central Toronto             1.0          Tennis Court   
91  Downtown Toronto             1.0                  Park   
98         Etobicoke             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop           Yoga Studio         Deli / Bodega   
16                        Park                 Field          Hockey Arena   
21               Women's Store                  Pool         Deli / Bodega   
35                        Park     Convenience Store           Yoga Studio   
49  Construction & Landscaping                Bakery           Yoga Studio   
52                 Yoga Studio         Deli / Bodega           Escape Room   
66           Convenience Store           Yoga Studio      Department Store   
83                        Park                 Trail        Discount Store   
91                  Playground                 Trail           Yoga Studio   
98                        Pool                 River           Yoga Studio   

   5th Most Common Venue        6th Most Common Venue  \
0      Electronics Store  Eastern European Restaurant   
16           Yoga Studio                 Dessert Shop   
21     Electronics Store  Eastern European Restaurant   
35         Deli / Bodega            Electronics Store   
49      Department Store                  Escape Room   
52     Electronics Store  Eastern European Restaurant   
66           Escape Room            Electronics Store   
83         Deli / Bodega             Department Store   
91          Dance Studio  Eastern European Restaurant   
98          Dance Studio  Eastern European Restaurant   

          7th Most Common Venue        8th Most Common Venue  \
0                     Drugstore                   Donut Shop   
16           Dim Sum Restaurant                        Diner   
21                    Drugstore                   Donut Shop   
35  Eastern European Restaurant                    Drugstore   
49            Electronics Store  Eastern European Restaurant   
52                    Drugstore                   Donut Shop   
66  Eastern European Restaurant                    Drugstore   
83                 Dessert Shop           Dim Sum Restaurant   
91                    Drugstore                   Donut Shop   
98                    Drugstore                   Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
0                Dog Run    Distribution Center  
16        Discount Store                Dog Run  
21               Dog Run    Distribution Center  
35            Donut Shop                Dog Run  
49             Drugstore             Donut Shop  
52               Dog Run    Distribution Center  
66            Donut Shop                Dog Run  
83                 Diner            Yoga Studio  
91               Dog Run    Distribution Center  
98               Dog Run    Distribution Center

Cluster 3

In [57]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6      Department Store           Escape Room     Electronics Store   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6  Eastern European Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6               Dog Run    Distribution Center

Cluster 4

In [58]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue 2nd Most Common Venue  \
32  Scarborough             3.0              Playground        Cosmetics Shop   
50   North York             3.0  Furniture / Home Store          Intersection   
85  Scarborough             3.0                    Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32           Yoga Studio         Deli / Bodega     Electronics Store   
50           Yoga Studio      Department Store           Escape Room   
85          Intersection           Yoga Studio         Deli / Bodega   

          6th Most Common Venue        7th Most Common Venue  \
32  Eastern European Restaurant                    Drugstore   
50            Electronics Store  Eastern European Restaurant   
85            Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32            Donut Shop               Dog Run    Distribution Center  
50             Drugstore            Donut Shop                Dog Run  
85             Drugstore            Donut Shop                Dog Run

Cluster 5

In [60]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
64    York             4.0         Jewelry Store           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
64      Department Store           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
64  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
64               Dog Run    Distribution Center